main.go

```diff
@@ -154,3 +220,197 @@ func gatekeeperNamespace(platformInfo platform.PlatformInfo) (string, error) {
+	}
+
+	auditFromCache := audit["auditFromCache"]
+	if auditFromCache != string(operatorv1alpha1.AuditFromCacheAutomatic) {
+		setupLog.V(1).Info("auditFromCache is not Automatic. ConstraintStatus controller cannot be created")
+		return false
+	}
+
+	// Start checking constraintpodstatuses.status.gatekeeper.sh CRD exist and available
+	constraintCrdGVR := schema.GroupVersionResource{
+		Group:    "apiextensions.k8s.io",
+		Version:  "v1",
+		Resource: "customresourcedefinitions",
+	}
+
+	crdUnstructured, err := dynamicClient.Resource(constraintCrdGVR).Get(mainCtx, "constraintpodstatuses.status.gatekeeper.sh",
+		metav1.GetOptions{})
+	if err != nil {
+		setupLog.V(1).Info("Parsing Constraintpodstatuses crd has error (crd), Wait for crd status update")
+		time.Sleep(time.Second)
+		return c.isApiAvailable(mainCtx, result, dynamicClient)
+	}
+
+	conditions, _, err := unstructured.NestedSlice(crdUnstructured.Object, "status", "conditions")
+	if err != nil {
+		setupLog.V(1).Info("Parsing Constraintpodstatuses crd has error (Condition), Wait for crd status update")
+		time.Sleep(time.Second)
+		return c.isApiAvailable(mainCtx, result, dynamicClient)
+	}
+
+	for _, condition := range conditions {
+		conditionObj, ok := condition.(map[string]interface{})
+		if !ok {
+			continue
+		}
+
+		conditionStatus, ok := conditionObj["status"].(string)
+		if !ok {
+			continue
+		}
+
+		conditionType, ok := conditionObj["type"].(string)
+		if !ok {
+			continue
+		}
+
+		if conditionType == string(v1.NamesAccepted) && conditionStatus == "True" {
+			return true
+		}
+	}
+
+	setupLog.V(1).Info("Constraintpodstatuses crd is not ready, Wait for crd status update")
+
+	time.Sleep(time.Second)
+	return c.isApiAvailable(mainCtx, result, dynamicClient)
+}
+
+// Watch the constraintpodstatuses resource in this function. When adding or Modified events comes,
+// This function adds constraintpodstatus function. Modified events are expected as status changes
+// When deleting event comes, close the function using context
+func (c *ConstraintStatusInstaller) watchGatekeeperInstall(mainCtx context.Context, dynamicClient *dynamic.DynamicClient,
+	cfg *rest.Config, managerOptions manager.Options, namespace string,
+) {
+	var ctx context.Context
+	var ctxCancel context.CancelFunc
+
+	watcher, _ := c.newWatcher(mainCtx, dynamicClient)
+	for {
+		select {
+		case result, ok := <-watcher.ResultChan():
+			// the channel got closed, so we need to restart
+			if !ok {
+				// main context closed. This loop should be closed.
+				if mainCtx.Err() != nil {
+					ctxCancel()
+					return
+				}
+				watcher, _ = c.newWatcher(mainCtx, dynamicClient)
+				continue
+			}
+
+			switch result.Type { //nolint:exhaustive
+			case watch.Modified, watch.Added:
+				if c.getIsInstalled() {
+					continue
+				}
+
+				// Check constraintpodstatuses crd available
+				if !c.isApiAvailable(mainCtx, result, dynamicClient) {
+					break
+				}
+
+				ctx, ctxCancel = context.WithCancel(mainCtx)
```

Comment:
It's probably better to just instantiate this at the top of the function so it's always defined.

Reply:
We need new a context after context canceled

Reply:
I'll rephrase what I said: I think you should instantiate it at the top of the function and then reassign it in the `case watch.Deleted:` case after the context is cancelled so that way there is a fresh context if `Gatekeeper` is instantiated again later.